lib/jxl/modular/transform/squeeze.cc

```diff
@@ -43,36 +185,105 @@ Status InvHSqueeze(Image &input, uint32_t c, uint32_t rc, ThreadPool *pool) {
     return true;
   }
 
+#if HWY_TARGET != HWY_SCALAR
+
+  // somewhat complicated trickery just to be able to SIMD this.
+  // Horizontal unsqueeze has horizontal data dependencies, so we do
+  // 8 rows at a time and treat it as a vertical unsqueeze of a
+  // transposed 8x8 block (or 9x8 for one input).
+  intptr_t onerow_in = chin.plane.PixelsPerRow();
+  intptr_t onerow_inr = chin_residual.plane.PixelsPerRow();
+  intptr_t onerow_out = chout.plane.PixelsPerRow();
+  constexpr size_t kRowsPerThread = 8;
   JXL_RETURN_IF_ERROR(RunOnPool(
-      pool, 0, chin.h, ThreadPool::NoInit,
+      pool, 0, chin.h / kRowsPerThread, ThreadPool::NoInit,
+      [&](const uint32_t task, size_t /* thread */) {
+        const size_t y0 = task * kRowsPerThread;
+        const pixel_type *JXL_RESTRICT p_residual = chin_residual.Row(y0);
+        const pixel_type *JXL_RESTRICT p_avg = chin.Row(y0);
+        pixel_type *JXL_RESTRICT p_out = chout.Row(y0);
+
+        HWY_ALIGN pixel_type b_p_avg[9 * kRowsPerThread];
+        HWY_ALIGN pixel_type b_p_residual[8 * kRowsPerThread];
+        HWY_ALIGN pixel_type b_p_out_even[kRowsPerThread];
+        HWY_ALIGN pixel_type b_p_out_odd[kRowsPerThread];
+        size_t x = 0;
+        if (chin_residual.w > 16) {
+          for (; x < chin_residual.w - 9; x += 8) {
+            Transpose8x8Block(p_residual + x, b_p_residual, onerow_inr);
+            Transpose8x8Block(p_avg + x, b_p_avg, onerow_in);
+            for (size_t y = 0; y < kRowsPerThread; y++) {
+              b_p_avg[8 * 8 + y] = p_avg[x + 8 + onerow_in * y];
+            }
+            for (size_t i = 0; i < 8; i++) {
+              fast_unsqueeze<kRowsPerThread>(
+                  b_p_residual + 8 * i, b_p_avg + 8 * i, b_p_avg + 8 * (i + 1),
+                  (x + i ? b_p_out_odd : b_p_avg + 8 * i), b_p_out_even,
+                  b_p_out_odd);
+              for (size_t y = 0; y < kRowsPerThread; y++) {
+                p_out[((x + i) << 1) + onerow_out * y] = b_p_out_even[y];
```

Comment:
Probably want to use StoreInterleaved from simd_util (and probably you want to do this directly in FastUnsqueeze, i.e. templatizing over a functor that takes the vector to be written, and transposes them and then writes them here or writes them directly in vertical squeeze)

Another option is to have FastUnsqueeze return a std::array<xxx, 8> (and perhaps making kRowsPerThread a global constant).

Reply:
It looks like at least with my clang, autovec seems to already figure it out.

Reply:
I'd still prefer to see it explicitly :)

Reply:
Done. Now the un-transpose and interleaved storing is made explicit.